### PR TITLE
Nesting of block indented expressions

### LINKED ISSUE #1422: Nesting of block indented expressions
See https://github.com/rust-lang-nursery/fmt-rfcs/issues/61

### LINKED ISSUE #1449: closures and block formatting very verbose
@ChristopherDavenport was helpfully applying rustfmt to rayon, using the [rfc-rustfmt](https://github.com/rust-lang-nursery/rustfmt/blob/master/rfc-rustfmt.toml) settings. The new fn signatures look great, but [closures result in quite a lot of indentation](https://github.com/nikomatsakis/rayon/pull/280/files#diff-6a14094077d75dad989b9e8f9a440937R103):

```rust
#[test]
fn future_rayon_wait_1_thread() {
    // run with only 1 worker thread; this would deadlock if we couldn't make progress
    let mut result = None;
    ThreadPool::new(Configuration::new().num_threads(1))
        .unwrap()
        .install(
            || {
                scope(
                    |s| {
                        use std::sync::mpsc::channel;
                        let (tx, rx) = channel();
                        let a = s.spawn_future(lazy(move || Ok::<usize, ()>(rx.recv().unwrap())));
                        //                          ^^^^ FIXME: why is this needed?
                        let b = s.spawn_future(a.map(|v| v + 1));
                        let c = s.spawn_future(b.map(|v| v + 1));
                        s.spawn(move |_| tx.send(20).unwrap());
                        result = Some(c.rayon_wait().unwrap());
                    },
                );
            },
        );
    assert_eq!(result, Some(22));
}
```

### LINKED ISSUE #1494: Special-case functions that take a single closure as an argument?
In bindgen we use a lot functions like `Option::or_else`.

When I rust `rustfmt` on bindgen with the rfc style (or something close-ish):

```toml
fn_args_layout = "Block"
array_layout = "Block"
where_style = "Rfc"
generics_indent = "Block"
fn_call_style = "Block"
```

I get diffs that look like this:

<details>

```diff
@@ -169,25 +161,28 @@ impl Cursor {
         // have to.
         self.cur_type()
             .num_template_args()
-            .or_else(|| {
-                let n: c_int =
-                    unsafe { clang_Cursor_getNumTemplateArguments(self.x) };
-
-                if n >= 0 {
-                    Some(n as u32)
-                } else {
-                    debug_assert_eq!(n, -1);
-                    None
-                }
-            })
-            .or_else(|| {
-                let canonical = self.canonical();
-                if canonical != *self {
-                    canonical.num_template_args()
-                } else {
-                    None
-                }
-            })
+            .or_else(
+                || {
+                    let n: c_int = unsafe { clang_Cursor_getNumTemplateArguments(self.x) };
+
+                    if n >= 0 {
+                        Some(n as u32)
+                    } else {
+                        debug_assert_eq!(n, -1);
+                        None
+                    }
+                },
+            )
+            .or_else(
+                || {
+                    let canonical = self.canonical();
+                    if canonical != *self {
+                        canonical.num_template_args()
+                    } else {
+                        None
+                    }
+                },
+            )
```

</details>

I'm not sure it's intended, I think it'd be nice to special-case functions taking only one closure as an argument, to avoid the double indentation and the trailing comma.

### PR DESCRIPTION
**Pull request by topecongiro**
Closes #1422, closes #1449, and closes #1494.

### CODE REVIEW

File: src/expr.rs [hunk 1]
```diff
@@ -1318,18 +1318,21 @@ impl Rewrite for ast::Arm {
...
-                    expr
+                    (false, &**expr)
                 } else {
-                    &**body
+                    (false, &**body)
                 }
             }
-            _ => &**body,
+            ast::ExprKind::Call(_, ref args) => (args.len() == 1, &**body),
+            ast::ExprKind::Closure(..) => (true, &**body),
```
Comment (nrc) [inline src/expr.rs]: Should we handle tuples here too?

Comment (topecongiro) [inline src/expr.rs]: Tuples are not refered in https://github.com/rust-lang-nursery/fmt-rfcs/issues/61, but if we should, do we use "a single expression parameter" rule or allow exception like clousres?
```rust
// A single expression tuple
let a = foo(bar((|x| {
    let y = x + 1;
    let z = y + 1;
    z
})));

// Should we allow this?
let a = foo(bar((b, c, |x| {
    let y = x + 1;
    let z = y + 1;
    z
})));
```

Comment (nrc) [inline src/expr.rs]: I think we should allow both (note that tuples with a single element must always have a trailing comma).

Comment (topecongiro) [inline src/expr.rs]: Thanks for clarifying!

Comment (topecongiro) [inline src/expr.rs]: I would like to make it clear, should we use block indent for tuples when `fn_call_style = Block` and format like function call? (c.f. https://github.com/rust-lang-nursery/fmt-rfcs/issues/64)
```rust
let a = (
    xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx,
    yyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyyy,
);
```

Comment (nrc) [inline src/expr.rs]: yes, I think so

**Comment by topecongiro** [comment]
I have added tuple and block. Also added a commit to add a trailing comma to a single arg in multiline.
Now rewriting tuple uses the same function as rewriting function call when `fn_call_style = Block`.

**Comment by nrc** [comment]
Awesome, thank you!

**Pull request merged**